Make map and array functions use IS DISTINCT semantics as appropriate

jirassimok opened the pull request:
This addresses #560.

Release note:
```markdown
# General
* Compare with "is distinct" semantics in `array_except`, `array_union`, `map_concat`, `map_from_entries`, `multimap_from_entries`, and `multimap_agg`. ({issue}`560`)
```

jirassimok commented:
I guess I should split the various refactors into a separate PR.

findepi commented:
> I've split the refactor commits into #9901; only the last four commits here are part of this PR.

thanks

jirassimok commented:
I'm not actually sure we should make this change for some of the Map functions (unless we also define Maps as using distinctness instead of equality for key lookups, which I don't think is an especially good idea).

Alternatively, we could make the map functions check for not-equal-but-not-distinct values and fail on them, which I think might be better (though it would make the functions a bit slower to maintain a second `TypedSet` for that).

(Right now, the only equal-but-not-distinct value I know of is NaN, and floating-point values shouldn't be used as map keys anyway because they don't do equality comparison well in the first place.)

findepi commented:
rebased after https://github.com/trinodb/trino/pull/9905 merged.

jirassimok commented:
@martint @dain What do you think?

dain commented:
This looks directionally correct and nicely simple :)

jirassimok commented:
Rebased and added a test for `multimap_agg`.

There still is no test for `MapToMapCast` because I can't think of a good way to test it—I don't know of any values that change distinctness when cast to another type.

findepi commented:
> I don't know of any values that change distinctness when cast to another type.

double to long  ( input 1.1 and 1.2)   or vice versa (input: certain consecutive values over 2^53)

jirassimok commented:
Casting to duplicate keys is already tested, so I guess I don't need to add a new test for that here.

jirassimok commented:
Rebased onto master.

jirassimok commented:
@martint Look good to you?

joshthoward commented:
@jirassimok I had an offline discussion with @dain and he said that while he thought it looked directionally correct he wouldn't have time to complete a full review. 

@findepi or @losipiuk could either of you provide a review to close this issue? 

CC: @martint

losipiuk commented on core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayExceptFunction.java:
```diff
@@ -66,4 +66,10 @@ public void testDuplicates()
+    @Test
+    public void testNonDistinctEqualValues()
+    {
+        assertFunction("array_except(ARRAY[NaN()], ARRAY[NaN()])", new ArrayType(DOUBLE), ImmutableList.of());
```
Mix with some normal numbers, so we do not end up with empty result which is kinda a corner case.

losipiuk commented on core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayExceptFunction.java:
```diff
@@ -66,4 +66,10 @@ public void testDuplicates()
     }
+
+    @Test
+    public void testNonDistinctEqualValues()
```
testNonDistinctNonEqualValues

losipiuk commented on core/trino-main/src/test/java/io/trino/type/TestArrayOperators.java:
```diff
@@ -1341,6 +1341,7 @@ public void testArrayUnion()
         assertFunction("ARRAY_UNION(ARRAY [8.3E0, 1.6E0, 4.1E0, 5.2E0], ARRAY [4.0E0, 5.2E0, 8.3E0, 9.7E0, 3.5E0])", new ArrayType(DOUBLE), ImmutableList.of(8.3, 1.6, 4.1, 5.2, 4.0, 9.7, 3.5));
         assertFunction("ARRAY_UNION(ARRAY [5.1E0, 7, 3.0E0, 4.8E0, 10], ARRAY [6.5E0, 10.0E0, 1.9E0, 5.1E0, 3.9E0, 4.8E0])", new ArrayType(DOUBLE), ImmutableList.of(5.1, 7.0, 3.0, 4.8, 10.0, 6.5, 1.9, 3.9));
         assertFunction("ARRAY_UNION(ARRAY [ARRAY [4, 5], ARRAY [6, 7]], ARRAY [ARRAY [4, 5], ARRAY [6, 8]])", new ArrayType(new ArrayType(INTEGER)), ImmutableList.of(ImmutableList.of(4, 5), ImmutableList.of(6, 7), ImmutableList.of(6, 8)));
+        assertFunction("ARRAY_UNION(ARRAY [NaN()], ARRAY [NaN()])", new ArrayType(DOUBLE), ImmutableList.of(NaN)); // results unique based on IS DISTINCT semantics
```
mix with some normal values.

losipiuk commented on core/trino-main/src/test/java/io/trino/operator/aggregation/TestMultimapAggAggregation.java:
```diff
@@ -88,6 +88,13 @@ public void testNullMap()
+    @Test
+    public void testKeysUseIsDistinctSemantics()
+    {
+        testMultimapAgg(DOUBLE, ImmutableList.of(Double.NaN, Double.NaN), BIGINT, ImmutableList.of(1L, 1L));
```
this testing sucks; we should explicity compare with expected multimap. Not construct it risking same bug in constructing and tested code. (no changes requested)

jirassimok replied:
Agreed.

losipiuk commented on core/trino-main/src/main/java/io/trino/operator/scalar/MapToMapCast.java:
```diff
@@ -238,14 +238,14 @@ public static Block mapCast(
         checkState(targetType.getTypeParameters().size() == 2, "Expect two type parameters for targetType");
         Type toKeyType = targetType.getTypeParameters().get(0);
-        TypedSet resultKeys = createEqualityTypedSet(toKeyType, keyEqual, keyHashCode, fromMap.getPositionCount() / 2, "map-to-map cast");
+        TypedSet resultKeys = createDistinctTypedSet(toKeyType, keyDistinctOperator, keyHashCode, fromMap.getPositionCount() / 2, "map-to-map cast");
```
You can somewhat test it with:
```
select cast(map(array['NaN', 'NaN '], array[1, 2]) as map<double, bigint>);
```

It will fail now. It used to not.

jirassimok replied:
Added a test.

losipiuk commented:
Code looks good for me. I would love to learn more about motivation. It is not obvious to me what semantics should be actually, hence why is this a fix for "correctness" issue.
Also: https://github.com/trinodb/trino/issues/560#issuecomment-1022455870

@findepi please comment.

findepi commented:
> Code looks good for me. I would love to learn more about motivation. It is not obvious to me what semantics should be actually, hence why is this a fix for "correctness" issue. Also: [#560 (comment)](https://github.com/trinodb/trino/issues/560#issuecomment-1022455870)
> 
> @findepi please comment.

that's an old conversion... this is @dain 's comment https://github.com/prestodb/presto/pull/11979#issuecomment-441804943
couple other references you can find @martint involvement there as well
https://github.com/prestodb/presto/issues/11963 https://github.com/prestodb/presto/issues/11978 https://github.com/prestodb/presto/pull/11979  https://github.com/prestodb/presto/issues/11977 https://github.com/prestodb/presto/pull/12080

jirassimok commented on core/trino-main/src/main/java/io/trino/operator/scalar/ArrayUnionFunction.java:
```diff
@@ -49,9 +48,9 @@ private ArrayUnionFunction() {}
             @TypeParameter("E") Type type,
             @OperatorDependency(
-                    operator = EQUAL,
+                    operator = IS_DISTINCT_FROM,
```
Easy fix.

jirassimok commented:
Rebased onto master.

losipiuk commented:
> @dain @losipiuk I believe this is ready to merge.

Yeah I think. @jirassimok can you update PR description to use tample (specifically putting release notes entry there) :)

jirassimok commented:
Added release note to PR description.

losipiuk commented:
rebased
